src/components/implementation/no_interface/llbooter/boot_deps.h

```diff
@@ -135,229 +189,391 @@ boot_newcomp_sinv_alloc(spdid_t spdid)
-	 * Get the first spdid that needs us to switch to it
-	 */
+	PRINTC("Booter: done creating system.\n");
+	PRINTC("********************************\n");
+	cos_thd_switch(schedule[sched_cur]);
+	PRINTC("Booter: done initializing child components.\n");
 
-	while (init_thds[global_idx] == 0) global_idx++;
-	assert(global_idx < INIT_THDS_SIZE);
-	cos_thd_switch(init_thds[global_idx]);
+	if (root_spdid) {
+		/* NOTE: Chronos delegations would replace this in some  experiments! */
+		root_aep = boot_spd_initaep_get(root_spdid);
 
-	/*
-	 * ERROR, we should not end back up in the llbooter's thread.
-	 * When the switched to component is done initializing, it will
-	 * invoke the booter in boot_thd_done where it will itself switch
-	 * to the next component that needs to be initialized.
-	 */
-	assert(0);
+		PRINTC("Root scheduler is %u, switching to it now!\n", root_spdid);
+		ret = cos_tcap_transfer(root_aep->rcv, BOOT_CAPTBL_SELF_INITTCAP_BASE, TCAP_RES_INF, LLBOOT_ROOT_PRIO);
+		assert(ret == 0);
+
+		ret = cos_switch(root_aep->thd, root_aep->tc, LLBOOT_ROOT_PRIO, TCAP_TIME_NIL, 0, cos_sched_sync());
+		PRINTC("ERROR: Root scheduler returned.\n");
+		assert(0);
+	}
+
+	PRINTC("No root scheduler in the system. Spinning!\n");
+	while (1) ;
 }
 
-static int
-boot_check_scheduler(char *comp_name)
+void
+boot_thd_done(spdid_t c)
 {
-	int ret;
-	const char prefix[] = "sl_";
+	struct comp_cap_info *acomp = &new_comp_cap_info[c];
 
-	ret = !strncmp(comp_name, prefix, sizeof(prefix)-1);
+	assert(acomp->parent_spdid == 0);
+	sched_cur++;
 
-	return ret;
+	PRINTC("Component %d initialized!\n", c);
+	if (schedule[sched_cur] != 0) {
+		cos_thd_switch(schedule[sched_cur]);
+	} else {
+		cos_thd_switch(BOOT_CAPTBL_SELF_INITTHD_BASE);
+	}
 }
 
-static int
-boot_check_shdmem(char *comp_name)
+static inline int
+boot_comp_initthd_get(spdid_t dstid, spdid_t srcid, thdcap_t thdslot, arcvcap_t rcvslot, tcap_t tcslot)
 {
-	int ret;
-	const char prefix[] = "shmem";
+	struct cos_compinfo  *boot_info = boot_spd_compinfo_get(0);
+	struct comp_cap_info *acomp     = &new_comp_cap_info[srcid];
+	struct cos_compinfo  *resci     = boot_spd_compinfo_get(dstid);
+	struct cos_aep_info  *initaep   = boot_spd_initaep_get(srcid);
+	int ret = 0;
 
-	ret = !strncmp(comp_name, prefix, sizeof(prefix)-1);
+	if (srcid > num_cobj || dstid > num_cobj) return -1;
+
+	if (initaep->thd) {
+		thdcap_t t;
+
+		ret = cos_cap_cpy_at(resci, thdslot, boot_info, initaep->thd);
+		assert(ret == 0);
+
+		if (acomp->is_sched) {
+			assert(initaep->rcv && initaep->tc);
+			ret = cos_cap_cpy_at(resci, rcvslot, boot_info, initaep->rcv);
+			assert(ret == 0);
+
+			ret = cos_cap_cpy_at(resci, tcslot, boot_info, initaep->tc);
+			assert(ret == 0);
+		}
+
+		return 0;
+	}
 
-	return ret;
+	return -1;
 }
 
-void
-boot_thd_done(void)
+static int 
+boot_comp_info_get(spdid_t dstid, spdid_t srcid, pgtblcap_t pgtslot, captblcap_t captslot, compcap_t ccslot, spdid_t *psid)
```

Comment:
`capt` isn't something that we've used.  `ct` or `captbl`

Don't know from the name what `ccslot` is.

Is `psid` a mistyped `spdid`?  If it is a parent scheduler, then it likely needs a more explicit name as no part of this function seems to be about scheduling...

Reply:
ccslot -> component capability slot. :-P

`psid` is parent spdid. I'll rename it to `par_id` or something.